mlir/lib/Dialect/Linalg/Transforms/Hoisting.cpp

```diff
@@ -152,6 +152,14 @@ void mlir::linalg::hoistRedundantVectorTransfers(func::FuncOp func) {
           transferRead.getPermutationMap() != transferWrite.getPermutationMap())
         return WalkResult::advance();
 
+      // When the source of transfer_read aliases, the following dominance
+      // analysis might not be sufficient.
+      // TODO: There might be other, similar cases missing here (i.e. other
+      // Memref Ops). 
+      auto source = transferRead.getSource();
+      if (source.getDefiningOp<memref::CollapseShapeOp>())
```

Comment:
This is a bit too adhoc to my taste, can we use `ViewLikeOpInterface` to capture all the cases that may alias?

From what I see we should also update noAliasingUseInLoop to use `ViewLikeOpInterface` and avoid enumerating an incomplete list of ops.

Lastly, you also want to add a bit more documentation to `hoistRedundantVectorTransfers` point 3. to mention that one may also want to fold aliasing memref operations into `vector.transfer` ops (in addition to LICM) to improve the hoisting.

Reply:
Makes sense, thanks a ton for the pointers! 

Patch updated. Hopefully I haven't missed/misunderstood anything.